#include "wood.h"
#include "binarywoods.h"
#include "ui_binarywoods.h"
#include <QIntValidator>
#include <QTime>
#include <QGraphicsView>
#include <QGraphicsScene>
#include <QGraphicsTextItem>
#include <cmath>
#define EMERGENCY_STOP "üõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõëüõë" // –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ—å–Ω–æ—Å—Ç—å, –∫–æ—Ç–æ—Ä—É—é —è–≤–Ω–æ –Ω–µ–ª—å–∑—è –≤–≤–µ—Å—Ç–∏, –Ω—É–∂–Ω–∞ –¥–ª—è –æ—Å—Ç–∞–Ω–æ–≤–∫–∏ –≤–≤–æ–¥–∞ —ç–ª–µ–º–µ–Ω—Ç–æ–≤

QString str = ""; // –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –≤–≤–æ–¥–∞
std::vector<std::pair<QString, QString>> sea; //–≤–µ–∫—Ç–æ—Ä –¥–ª—è –ø–æ–∏—Å–∫–∞: 1 - –ø—É—Ç—å, 2 - —Å–∞–º —ç–ª–µ–º–µ–Ω—Ç
std::vector<QString> obh; // –≤–µ–∫—Ç–æ—Ä –¥–ª—è –æ–±—Ö–æ–¥–æ–≤

BinaryWoods::BinaryWoods(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::BinaryWoods)
{
    ui->setupUi(this);
    connect(ui->createWoodBTN, &QPushButton::clicked, this, &BinaryWoods::createWood); //–∫–æ–Ω–Ω–µ–∫—Ç—ã –∫–Ω–æ–ø–æ–∫
    connect(ui->okBTN, &QPushButton::clicked, this, &BinaryWoods::getText);
    connect(ui->balanceBTN, &QPushButton::clicked, this, &BinaryWoods::balanceTree);
    connect(ui->addElemsBTN, &QPushButton::clicked, this, &BinaryWoods::addElement);
    connect(ui->deleteElemBTN, &QPushButton::clicked, this, &BinaryWoods::delElement);
    connect(ui->delWoodBTN, &QPushButton::clicked, this, &BinaryWoods::delWood);
    connect(ui->findElemsBTN, &QPushButton::clicked, this, &BinaryWoods::searcher);
    connect(ui->closeSearchBTN, &QPushButton::clicked, this, &BinaryWoods::closeSearch);
    connect(ui->stopBTN, &QPushButton::clicked, this, &BinaryWoods::stop);
    connect(ui->directBTN, &QPushButton::clicked, this, &BinaryWoods::directObhod);
    connect(ui->closeObhodBTN, &QPushButton::clicked, this, &BinaryWoods::closeObhod);
    connect(ui->symetricalBTN, &QPushButton::clicked, this, &BinaryWoods::symmetricalObhod);
    connect(ui->reverseBTN, &QPushButton::clicked, this, &BinaryWoods::reverseObhod);

    ui->closeSearchBTN->setHidden(true);//–ø—Ä—è—á–µ–º –ø–æ–∫–∞ –Ω–µ –Ω—É–∂–Ω—ã–µ –∫–Ω–æ–ø–∫–∏
    ui->stopBTN->setHidden(true);
    ui->closeObhodBTN->setHidden(true);

    graphicsView = ui->woodTV; //–∫–æ–Ω–Ω–µ–∫—Ç—ã –¥–µ—Ä–µ–≤—è–Ω–Ω–æ–≥–æ —Ç–µ–ª–µ–≤–∏–∑–æ—Ä–∞ –∏ —Å—Ü–µ–Ω—ã
    scene = new QGraphicsScene;
    graphicsView -> setScene(scene);

    srand(time(NULL)); // –æ–±–Ω–∞–≤–ª–µ–Ω–∏–µ —Å–∏–¥–∞ –≥–µ–Ω–µ—Ä–∞—Ç–æ—Ä–∞ —Å–ª—É—á–∞–π–Ω—ã—Ö —á–∏—Å–µ–ª
}

void delay() //–∂–¥—ë–º
{
    QTime dieTime= QTime::currentTime().addMSecs(1);
    while (QTime::currentTime() < dieTime) QCoreApplication::processEvents(QEventLoop::AllEvents, 100);
}

QString randomString() // –≥–µ–Ω–µ—Ä–∞—Ç–æ—Ä —Å–ª—É—á–∞–π–Ω–æ–π —Å—Ç—Ä–æ–∫–∏
{
    QString mass = "~!@#$%^&*()_+|<>{}[]`=-‚Ññ.,;:1234567890–ê–ë–í–ì–î–ï–Å–ñ–ó–ò–ô–ö–õ–ú–ù–û–ü–†–°–¢–£–§–•–¶–ß–®–©–™–´–¨–≠–Æ–Ø–∞–±–≤–≥–¥–µ—ë–∂–∑–∏–π–∫–ª–º–Ω–æ–ø—Ä—Å—Ç—É—Ñ—Ö—Ü—á—à—â—ä—ã—å—ç—é—èABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"; // –º–∞—Å—Å–∏–≤ —Å–∏–º–≤–æ–ª–æ–≤, –∏–∑ –∫–æ—Ç–æ—Ä—ã—Ö –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç—Å—è —Å–ª—É—á–∞–π–Ω–∞—è —Å—Ç—Ä–æ–∫–∞
    QChar qch; // —Å–∏–º–≤–æ–ª –¥–ª—è –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏
    QString s; // —Å—Ç—Ä–æ–∫–∞-—Ä–µ–∑—É–ª—å—Ç–∞—Ç
    int len = rand() % 16; // –≤—ã–±–∏—Ä–∞–µ–º –¥–ª–∏–Ω—É —Å—Ç—Ä–æ–∫–∏
    if (len==0) len = 1; // –ø—É—Å—Ç—ã–µ —Å—Ç—Ä–æ–∫–∏ –¥–µ–ª–∞–µ–º –¥–ª–∏–Ω–æ–π 1
    for(int i = 0; i < len; i++) // –∑–∞–ø–æ–ª–Ω—è–µ–º —Å—Ç—Ä–æ–∫—É
    {
        qch = mass[(rand() % mass.size())]; // –ø–æ–ª—É—á–∞–µ–º —Å–ª—É—á–∞–π–Ω—ã–π —Å–∏–º–≤–æ–ª
        s.push_back(qch); // –¥–æ–±–∞–≤–ª—è–µ–º –≤ —Å—Ç—Ä–æ–∫—É
    }
    return s; // –≤–æ–∑–≤—Ä–∞—â–∞–µ–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç
}

int maxDepth(Wood* T) //–Ω–∞—Ö–æ–¥–∏–º –≥–ª—É–±–∏–Ω—É —Ä–µ–∫—É—Ä—Å–∏–≤–Ω–æ
{
    if (T==NULL) return 0;
    return std::max(maxDepth(T->left),maxDepth(T->right))+1;
}

void BinaryWoods::allBTNsBlock() //–≤—ã–∫–ª—é—á–∞–µ–º –≤—Å–µ –∫–Ω–æ–ø–∫–∏
{
    ui->addElemsBTN->setEnabled(false);
    ui->createWoodBTN->setEnabled(false);
    ui->balanceBTN->setEnabled(false);
    ui->deleteElemBTN->setEnabled(false);
    ui->delWoodBTN->setEnabled(false);
    ui->findElemsBTN->setEnabled(false);
    ui->directBTN->setEnabled(false);
    ui->symetricalBTN->setEnabled(false);
    ui->reverseBTN->setEnabled(false);
}
void BinaryWoods::allBTNsUnBlock() //–≤–∫–ª—é—á–∞–µ–º –≤—Å–µ –∫–Ω–æ–ø–∫–∏
{
    ui->addElemsBTN->setEnabled(true);
    ui->createWoodBTN->setEnabled(true);
    ui->balanceBTN->setEnabled(true);
    ui->deleteElemBTN->setEnabled(true);
    ui->delWoodBTN->setEnabled(true);
    ui->findElemsBTN->setEnabled(true);
    ui->directBTN->setEnabled(true);
    ui->symetricalBTN->setEnabled(true);
    ui->reverseBTN->setEnabled(true);
}

void BinaryWoods::reCreateTV() //–ø–µ—Ä–µ—Å–æ–∑–¥–∞—Ç—å —Å—Ü–µ–Ω—É –∏ —Ç–µ–ª–µ–≤–∏–∑–æ—Ä, –æ—á–∏—â–∞–µ—Ç –≤—Å–µ –ø–æ–ª–Ω–æ—Å—Ç—å—é
{
    QGraphicsScene* del = scene;
    scene = new QGraphicsScene;
    delete del;
    ui->woodTV->setScene(scene);
    ui->woodTV->update();
}

void NewTree(Wood*&r, QChar* data, int masLen)   //—Ñ—É–Ω–∫—Ü–∏—è —Å–æ–∑–¥–∞–Ω–∏—è –Ω–æ–≤–æ–≥–æ –¥–µ—Ä–µ–≤–∞/–¥–æ–±–∞–≤–ª–µ–Ω–∏—è —ç–ª–µ–º–µ–Ω—Ç–∞
{
    if(NULL==r)
    {
        r=new Wood;
        r->left=r->right=NULL;
        r->data=data;       //–∑–∞–ø–∏—Å—å —ç–ª–µ–º–µ–Ω—Ç–∞
        r->dataSize = masLen; // –∏ –¥–ª–∏–Ω—ã –º–∞—Å—Å–∏–≤–∞
    }
    else
    {
        if(data<r->data)     //–∑–∞–ø–∏—Å—å –≤ –ª–µ–≤–æ–µ –ø–æ–¥–¥–µ—Ä–µ–≤–æ
        {
            if(r->left!=NULL) NewTree(r->left,data, masLen);
            else
            {
                r->left=new Wood;
                r->left->left=r->left->right=NULL;
                r->left->data=data;
                r->left->dataSize = masLen;
            }
        }
        else                 //–∑–∞–ø–∏—Å—å –≤ –ø—Ä–∞–≤–æ–µ –ø–æ–¥–¥–µ—Ä–µ–≤–æ
        {
            if(r->right!=NULL) NewTree(r->right,data, masLen);
            else
            {
                r->right=new Wood;
                r->right->right=r->right->left=NULL;
                r->right->data=data;
                r->right->dataSize = masLen;
            }
        }
    }
    r->depth = maxDepth(r); // –∑–∞–ø–∏—Å—å –≥–ª—É–±–∏–Ω—ã
}

void collectNodes(Wood* node, std::vector<std::pair<QChar*, int>>& nodes) //–ø–æ–ª—É—á–∞–µ—Ç –≤–µ–∫—Ç–æ—Ä –ø–∞—Ä, –≥–¥–µ –ø–µ—Ä–≤–æ–µ - –¥–∞–Ω–Ω—ã–µ —É–∑–ª–∞ –¥–µ—Ä–µ–≤–∞ (–º–∞—Å—Å–∏–≤ —Å–∏–º–≤–æ–ª–æ–≤), –≤—Ç–æ—Ä–æ–µ - –µ–≥–æ —Ä–∞–∑–º–µ—Ä
    {
        if (node == nullptr)
        {
            return;
        }
        collectNodes(node->left, nodes);
        collectNodes(node->right, nodes);
        nodes.push_back(std::make_pair(node->data, node->dataSize));
    }

Wood* buildBalancedTree(const std::vector<std::pair<QChar*, int>>& nodes, int start, int finish) //—Å–æ–∑–¥–∞–µ—Ç —Å–±–∞–ª–∞–Ω—Å–∏—Ä–æ–≤–∞–Ω–Ω–æ–µ –¥–µ—Ä–µ–≤–æ –∏–∑ –≤–µ–∫—Ç–æ—Ä–∞ –ø–∞—Ä
{
        if (start > finish)
        {
            return nullptr;
        }

        int mid = (start + finish) / 2;
        Wood* newNode = new Wood;
        newNode->data = nodes[mid].first;
        newNode->dataSize = nodes[mid].second;
        newNode->left = buildBalancedTree(nodes, start, mid - 1);
        newNode->right = buildBalancedTree(nodes, mid + 1, finish);
        return newNode;
}

void BinaryWoods::balanceTree() //—Å–æ–∑–¥–∞–µ—Ç —Å–±–∞–ª–∞–Ω—Å–∏—Ä–æ–≤–∞–Ω–Ω–æ–µ –¥–µ—Ä–µ–≤–æ –∏ –æ–±–Ω–æ–≤–ª—è–µ—Ç –¥–∏—Å–ø–ª–µ–π
{
        std::vector<std::pair<QChar*, int>> nodes;
        collectNodes(binaryWood, nodes);
        binaryWood = nullptr;
        std::sort(nodes.begin(), nodes.end());
        binaryWood = buildBalancedTree(nodes, 0, nodes.size() - 1);
        reCreateTV();
        binaryWood->depth = maxDepth(binaryWood);
        printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood);
}

void BinaryWoods::printWood(int sdvig, int x, int y, Wood* woo) //–≤—ã–≤–æ–¥–∏—Ç –¥–µ—Ä–µ–≤–æ –Ω–∞ –¥–µ—Ä–µ–≤—è–Ω–Ω—ã–π —Ç–µ–ª–µ–≤–∏–∑–æ—Ä —Ä–µ–∫—É—Ä—Å–∏–≤–Ω–æ
{
    if (binaryWood != nullptr)
    {
        QString text;
        for(int i = 0; i < woo->dataSize; i++) text.push_back(woo->data[i]);               // –≤—ã–≤–æ–¥ —Ç–µ–∫—É—â–µ–≥–æ —É–∑–ª–∞
        scene->addPolygon(QRectF(x, y, 200, 30), QPen(Qt::black), QBrush(Qt::lightGray));
        QGraphicsTextItem *textItem = scene->addText(text);
        textItem->setFont(QFont("Cascadia Code", 12, 2, false));
        textItem->setPos(x+4, y-2);
        if (woo->left != nullptr) // –≤—ã–≤–æ–¥ –ª–µ–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
        {
            printWood(sdvig/2, x-std::abs(sdvig), y+100, woo->left);
            scene->addLine(x+100, y+30, x-std::abs(sdvig)+100, y+100, QPen(Qt::black));
        }
        if (woo->right != nullptr) // –≤—ã–≤–æ–¥ –ø—Ä–∞–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
        {
            printWood(sdvig/2, x+std::abs(sdvig), y+100, woo->right);
            scene->addLine(x+100, y+30, x+std::abs(sdvig)+100, y+100, QPen(Qt::black));
        }
        ui->woodTV->update();
    }
}

void BinaryWoods::getText() //–ø–æ–ª—É—á–∞–µ—Ç —Ç–µ–∫—Å—Ç –≤ –≥–ª–æ–±–∞–ª—å–Ω—É—é —Å—Ç—Ä–æ–∫—É –ø–æ –Ω–∞–∂–∞—Ç–∏—é –û–ö
{
    if(ui->inputVsego->text()=="") return; // –µ—Å–ª–∏ —Ç–µ–∫—Å—Ç–∞ –Ω–µ—Ç, —Ç–æ –≤—ã—Ö–æ–¥–∏–º
    str = ui->inputVsego->text(); // –µ—Å–ª–∏ –µ—Å—Ç—å —Ç–æ –±–µ—Ä–µ–º
    ui->okBTN->setEnabled(false); // –≤—ã–∫–ª—é—á–∞–º –∫–Ω–æ–ø–∫—É
    ui->inputVsego->clear(); // —á–∏—Å—Ç–∏–º –ø–æ–ª–µ –≤–≤–æ–¥–∞
    ui->inputVsego->setEnabled(false); // –≤—ã–∫–ª—é—á–∞–µ–º –µ–≥–æ
    ui->textovoePole->setText(""); // —É–±–∏—Ä–∞–µ–º –∏–∑ –Ω–µ–≥–æ —Ç–µ–∫—Å—Ç
}

void BinaryWoods::stop() // –æ—Å—Ç–∞–Ω–æ–≤–∫–∞ –≤–≤–æ–¥–∞
{
    str = EMERGENCY_STOP; // –≤–≤–æ–¥–∏—Ç —Å–ø–µ—Ü–∏–∞–ª—å–Ω—É—é –ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ—Å—Ç—å, –æ—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞—é—â—É—é –≤–≤–æ–¥
}

void BinaryWoods::createWood() //—Å–æ–∑–¥–∞—Ç—å/–ø–µ—Ä–µ—Å–æ–∑–¥–∞—Ç—å –¥–µ—Ä–µ–≤–æ
{
    if (binaryWood !=nullptr) // –µ—Å–ª–∏ –µ—Å—Ç—å –¥–µ—Ä–µ–≤–æ - —É–¥–∞–ª–∏—Ç—å –¥–µ—Å—Ç—Ä—É–∫—Ç–æ—Ä —Å–ø—Ä–∞–≤–∏—Ç—Å—è, —è –ø—Ä–æ–≤–µ—Ä–∏–ª
    {
        binaryWood->~Wood();
        binaryWood = nullptr;
    }
    reCreateTV(); // –ø—Ä–æ—á–∏—Å—Ç–∏—Ç—å —Å—Ü–µ–Ω—É –∏ —Ç–µ–ª–∏–∫
    allBTNsBlock(); //–≤—ã–∫–ª –∫–Ω–æ–ø–∫–∏
    ui->textovoePole->setText("–í–≤–µ–¥–∏—Ç–µ –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ —ç–ª–µ–º–µ–Ω—Ç–æ–≤:");
    ui->inputVsego->setEnabled(true); //–¥–µ–ª–∞–µ–º –∞–∫—Ç–∏–≤–Ω—ã–º —Ç–æ —á—Ç–æ –Ω—É–∂–Ω–æ –∏ –º–µ–Ω—è–µ–º —Ç–µ–∫—Å—Ç
    ui->okBTN->setEnabled(true);
    QValidator *validator = new QIntValidator(1, 1000000000, this); // –æ–≥—Ä–∞–Ω–∏—á–∏–º –≤–≤–æ–¥
    ui->inputVsego->setValidator(validator);
    ui->inputVsego->setFocus();
    while(str=="") //–∂–¥—ë–º –≤–≤–æ–¥–∞
    {
        delay();
    }
    int elemKolvo = str.toInt();
    str = "";
    for(int i = 0; i < elemKolvo; i++) //–≤–≤–æ–¥–∏–º —Å–∞–º–∏ —ç–ª–µ–º–µ–Ω—Ç—ã
    {
        QString tekst = "–í–≤–µ–¥–∏—Ç–µ —ç–ª–µ–º–µ–Ω—Ç ‚Ññ";
        tekst.push_back(QString::number (i)); //–º–µ–Ω—è–µ–º —Ç–µ–∫—Å—Ç
        tekst.push_back(':');
        ui->textovoePole->setText(tekst);
        ui->inputVsego->setEnabled(true);// –¥–µ–ª–∞–µ–º –∞–∫—Ç–∏–≤–Ω—ã–º —Ç–æ–ª—å–∫–æ –Ω—É–∂–Ω–æ–µ
        ui->okBTN->setEnabled(true);
        ui->inputVsego->setMaxLength(15); // –æ–≥—Ä–∞–Ω–∏—á–∏–ª–∏ –≤–≤–æ–¥ –Ω–∞ 15 —Å–∏–º–≤–æ–ª–æ–≤
        ui->inputVsego->setValidator(NULL); // —É–±—Ä–∞–ª–∏ –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –Ω–∞ —á–∏—Å–ª–∞
        ui->inputVsego->setText(randomString()); // —Ä–∞–Ω–¥–æ–º–Ω–∞—è —Å—Ç—Ä–æ–∫–∞ –¥–ª—è –±—ã—Å—Ç—Ä–æ–≥–æ —Å–æ–∑–¥–∞–Ω–∏—è –¥–µ—Ä–µ–≤–∞
        ui->inputVsego->setSelection(0, 16); // –µ—Å–ª–∏ –≤–≤–µ–¥–µ–º —Å–≤–æ—é, —Ä–∞–Ω–¥–æ–º–Ω–∞—è —Å–æ—Ç—Ä–µ—Ç—Å—è, —Ç–∫ –≤—ã–¥–µ–ª–µ–Ω–∞
        ui->inputVsego->setFocus();
        ui->stopBTN->setHidden(false);

        while(str=="") //–∂–¥–µ–º –≤–≤–æ–¥–∞
        {
            delay();
        }

        if (str == EMERGENCY_STOP) // –µ—Å–ª–∏ –Ω–∞–∂–∞–ª–∏ —Å—Ç–æ–ø —Ç–æ –æ—Å—Ç–∞–Ω–æ–≤–∏—Ç—å —Ü–∏–∫–ª –∞–∫–∫—É—Ä–∞—Ç–Ω–æ
        {
            str="";
            ui->stopBTN->setHidden(true);
            ui->textovoePole->setText("");
            ui->inputVsego->clear();
            ui->inputVsego->setEnabled(false);
            ui->okBTN->setEnabled(false);
            break;
        }

        QChar* data = new QChar[str.size()]; //–ø–æ–ª—É—á–∏–ª–∏ –ø–æ–ª–µ –¥–∞–Ω–Ω—ã—Ö –¥–ª—è —É–∑–ª–∞ –¥–µ—Ä–µ–≤–∞
        for(int j = 0; j < str.size(); j++)
        {
            data[j] = str[j];
        }
        NewTree(binaryWood, data, str.size()); //–¥–æ–±–∞–≤–∏–ª–∏ —ç–ª–µ–º–µ–Ω—Ç –∏–ª–∏ —Å–æ–∑–¥–∞–ª–∏ –¥–µ—Ä–µ–≤–æ, –µ—Å–ª–∏ –æ–Ω –ø–µ—Ä–≤—ã–π
        str = ""; // –æ—á–∏—Å—Ç–∏–ª–∏ —Å—Ç—Ä–æ–∫—É
    }
    ui->stopBTN->setHidden(true); // —Å–ø—Ä—è—Ç–∞–ª–∏ –∫–Ω–æ–ø–∫—É —Å—Ç–æ–ø
    printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood); //–≤—ã–≤–µ–ª–∏ –¥–µ—Ä–µ–≤–æ –Ω–∞ –¥–µ—Ä–µ–≤—è–Ω–Ω—ã–π —Ç–µ–ª–µ–≤–∏–∑–æ—Ä
    ui->createWoodBTN->setText("–ü–µ—Ä–µ—Å–æ–∑–¥–∞—Ç—å –¥–µ—Ä–µ–≤–æ"); //–∏–∑–º–µ–Ω–∏–ª–∏ —Ç–µ–∫—Å—Ç –∫–Ω–æ–ø–∫–∏
    allBTNsUnBlock(); //–≤–∫–ª –∫–Ω–æ–ø–∫–∏
}

void BinaryWoods::addElement() // –¥–æ–±–∞–≤–∏—Ç—å —ç–ª–µ–º–µ–Ω—Ç
{
    allBTNsBlock(); //–≤—ã–∫–ª –∫–Ω–æ–ø–∫–∏
    QString tekst = "–í–≤–µ–¥–∏—Ç–µ —ç–ª–µ–º–µ–Ω—Ç:";
    ui->textovoePole->setText(tekst);
    ui->inputVsego->setEnabled(true);// –¥–µ–ª–∞–µ–º –∞–∫—Ç–∏–≤–Ω—ã–º —Ç–æ–ª—å–∫–æ –Ω—É–∂–Ω–æ–µ
    ui->okBTN->setEnabled(true);
    ui->inputVsego->setMaxLength(15); // –æ–≥—Ä–∞–Ω–∏—á–∏–ª–∏ –≤–≤–æ–¥ –Ω–∞ 15 —Å–∏–º–≤–æ–ª–æ–≤
    ui->inputVsego->setValidator(NULL); // —É–±—Ä–∞–ª–∏ –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –Ω–∞ —á–∏—Å–ª–∞
    ui->inputVsego->setFocus();
    while(str=="") //–∂–¥–µ–º –≤–≤–æ–¥–∞
    {
        delay();
    }
    QChar* data = new QChar[str.size()]; //–ø–æ–ª—É—á–∏–ª–∏ –ø–æ–ª–µ –¥–∞–Ω–Ω—ã—Ö –¥–ª—è —É–∑–ª–∞ –¥–µ—Ä–µ–≤–∞
    for(int j = 0; j < str.size(); j++)
    {
        data[j] = str[j];
    }
    NewTree(binaryWood, data, str.size()); //–¥–æ–±–∞–≤–∏–ª–∏ —ç–ª–µ–º–µ–Ω—Ç –∏–ª–∏ —Å–æ–∑–¥–∞–ª–∏ –¥–µ—Ä–µ–≤–æ, –µ—Å–ª–∏ –æ–Ω –ø–µ—Ä–≤—ã–π
    str = ""; // –æ—á–∏—Å—Ç–∏–ª–∏ —Å—Ç—Ä–æ–∫—É
    reCreateTV(); // –ø—Ä–æ—á–∏—Å—Ç–∏–ª–∏ —Å—Ü–µ–Ω—É –∏ —Ç–µ–ª–µ–∫
    binaryWood->depth = maxDepth(binaryWood); // –Ω–∞—à–ª–∏ –Ω–æ–≤—É—é –≥–ª—É–±–∏–Ω—É
    printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood); // –≤—ã–≤–µ–ª–∏ –¥–µ—Ä–µ–≤–æ
    allBTNsUnBlock(); //–≤–∫–ª –∫–Ω–æ–ø–∫–∏
}

Wood* BinaryWoods::pathWood(QString path) //–¥–æ—Ö–æ–¥–∏—Ç –ø–æ –ø—É—Ç–∏ –¥–æ –ø—Ä–µ–¥–ø–æ—Å–ª–µ–¥–Ω–µ–≥–æ —ç–ª–µ–º–µ–Ω—Ç–∞, –ø—Ä–æ–≤–µ—Ä—è—è –ø—Ä–∞–≤–∏–ª—å–Ω–æ—Å—Ç—å –≤—Å–µ–≥–æ –ø—É—Ç–∏
{
    Wood* current = binaryWood;
    for (int i = 0; i <path.size()-1; i++)
    {
        if (path[i] == 'r')
        {
            if (current->right != nullptr) current = current->right;
            else return nullptr; //—Ç–∫ –ø—É—Ç—å —É–∂–µ –Ω–µ–≤–µ—Ä–Ω—ã–π
        }
        else if (path[i] == 'l')
        {
            if (current->left != nullptr) current = current->left;
            else return nullptr; //—Ç–∫ –ø—É—Ç—å —É–∂–µ –Ω–µ–≤–µ—Ä–Ω—ã–π
        }
    }
    if ((path[path.size()-1] == 'r')&&(current->right) != nullptr) return current; //–ø—É—Ç—å –≤–µ—Ä–Ω—ã–π, –≤–æ–∑–≤—Ä–∞—Ç –ø—Ä–µ–¥–ø–æ—Å–ª. —ç–ª–µ–º–µ–Ω—Ç–∞
    if ((path[path.size()-1] == 'l')&&(current->left) != nullptr) return current; // –ø—É—Ç—å –≤–µ—Ä–Ω—ã–π, –≤–æ–∑–≤—Ä–∞—Ç –ø—Ä–µ–¥–ø–æ—Å–ª. —ç–ª–µ–º–µ–Ω—Ç–∞
    return nullptr; //—Ç–∫ –ø–æ—Å–ª–µ–¥–Ω–µ–µ –∑–≤–µ–Ω–æ –ø—É—Ç–∏ –Ω–µ–≤–µ—Ä–Ω–æ
}

void BinaryWoods::delElement() // —É–¥–∞–ª—è–µ—Ç —ç–ª–µ–º–µ–Ω—Ç, –∏ –∫–∞–∫ —Å–ª–µ–¥—Å—Ç–≤–∏–µ, –≤—Å—é –µ–≥–æ –≤–µ—Ç–∫—É
{
    allBTNsBlock(); //–≤—ã–∫–ª –∫–Ω–æ–ø–∫–∏
    QString tekst = "–í–≤–µ–¥–∏—Ç–µ –ø—É—Ç—å –≤ –≤–∏–¥–µ rrllrl:";
    ui->textovoePole->setText(tekst);
    ui->inputVsego->setEnabled(true);// –¥–µ–ª–∞–µ–º –∞–∫—Ç–∏–≤–Ω—ã–º —Ç–æ–ª—å–∫–æ –Ω—É–∂–Ω–æ–µ
    ui->okBTN->setEnabled(true);
    ui->inputVsego->setMaxLength(1000000); // –ø–æ—á—Ç–∏ —Å–Ω—è–ª–∏ –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –≤–≤–æ–¥–∞
    QRegExp re("[rl]+$"); // –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –Ω–∞ –±—É–∫–≤—ã
    QRegExpValidator*v = new QRegExpValidator(re);
    ui->inputVsego->setValidator(v); // —É–±—Ä–∞–ª–∏ –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –Ω–∞ —á–∏—Å–ª–∞ –∏ –ø–æ—Å—Ç–∞–≤–∏–ª–∏ –Ω–∞ –±—É–∫–≤—ã r l
    ui->inputVsego->setFocus();
    while(str=="") //–∂–¥–µ–º –≤–≤–æ–¥–∞
    {
        delay();
    }
    if(pathWood(str) != nullptr) //–µ—Å–ª–∏ –ø—É—Ç—å –≤–µ—Ä–Ω—ã–π
    {
        Wood* parentOfDelWood = pathWood(str); //–ø–æ–ª—É—á–∏–ª–∏ –ø—É—Ç—å –∏ –ø–æ –ø—É—Ç–∏ —É–∫–∞–∑–∞—Ç–µ–ª—å –Ω–∞ —Ä–æ–¥–∏—Ç–µ–ª—è —É–¥–∞–ª—è–µ–º–æ–≥–æ —ç–ª–µ–º–µ–Ω—Ç–∞
        Wood* delWood = (str[str.size()-1] == 'r' ? parentOfDelWood->right : parentOfDelWood->left);
        if (str[str.size()-1] == 'r') parentOfDelWood->right = nullptr;
        else parentOfDelWood->left = nullptr;
        delWood->~Wood(); //—É–¥–∞–ª–∏–ª–∏ –ø–æ–¥–¥–µ—Ä–µ–≤–æ
        reCreateTV();
        binaryWood->depth = maxDepth(binaryWood);
        printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood);
    }
    else
    {
        ui->textovoePole->setText("–ù–µ–≤–µ—Ä–Ω—ã–π –ø—É—Ç—å!");
        QTime dieTime= QTime::currentTime().addMSecs(1500);
        while (QTime::currentTime() < dieTime) QCoreApplication::processEvents(QEventLoop::AllEvents, 100);
        ui->textovoePole->setText("");
    }
    str = ""; // –æ—á–∏—Å—Ç–∏–ª–∏ —Å—Ç—Ä–æ–∫—É
    allBTNsUnBlock(); //–≤–∫–ª –∫–Ω–æ–ø–∫–∏
}

void pathSearch(QChar key, QString path, Wood*w) // –ø–æ–∏—Å–∫ —ç–ª–µ–º–Ω—Ç–∞ –ø–æ –ø–µ—Ä–≤–æ–º—É —Å–∏–º–≤–æ–ª—É —Ä–µ–∫—É—Ä—Å–∏–≤–Ω–æ, –≤ –≥–ª–æ–±–∞–ª—å–Ω—ã–π –≤–µ–∫—Ç–æ—Ä –¥–æ–±–∞–≤–ª—è–µ—Ç –ø–∞—Ä—ã –∏–∑ –Ω–∞–π–¥–µ–Ω–Ω–æ–≥–æ —ç–ª–µ–º–µ–Ω—Ç–∞ –∏ –µ–≥–æ –ø—É—Ç–∏
{
    if (w == nullptr) return;
    if (w->left != nullptr) pathSearch(key, path+"l", w->left);
    if (w->right != nullptr) pathSearch(key, path+"r", w->right);
    if (w->data[0] == key)
    {
        QString convertedData;
        for (int i = 0; i < w->dataSize; i++) convertedData.push_back(w->data[i]);
        sea.push_back(std::make_pair(path, convertedData));
    }
}

void BinaryWoods::searcher() // –ø–æ–∏—Å–∫ –∏ –≤—ã–≤–æ–¥ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤ –Ω–∞ —ç–∫—Ä–∞–Ω
{
    allBTNsBlock(); //–≤—ã–∫–ª –∫–Ω–æ–ø–∫–∏
    QString tekst = "–í–≤–µ–¥–∏—Ç–µ —Å–∏–º–≤–æ–ª –¥–ª—è –ø–æ–∏—Å–∫–∞:";
    ui->textovoePole->setText(tekst);
    ui->inputVsego->setEnabled(true); // –¥–µ–ª–∞–µ–º –∞–∫—Ç–∏–≤–Ω—ã–º —Ç–æ–ª—å–∫–æ –Ω—É–∂–Ω–æ–µ
    ui->okBTN->setEnabled(true);
    ui->inputVsego->setMaxLength(1); // –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –≤–≤–æ–¥–∞
    ui->inputVsego->setValidator(NULL); // —É–±—Ä–∞–ª–∏ –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏–µ –Ω–∞ —á–∏—Å–ª–∞
    ui->inputVsego->setFocus();
    while(str=="") //–∂–¥–µ–º –≤–≤–æ–¥–∞
    {
        delay();
    }
    QChar key = str[0];
    str="";
    pathSearch(key, "", binaryWood);

    int y = 0;
    for(unsigned int i = 0; i < sea.size(); i++) // –ø–µ—á–∞—Ç–∞–µ–º –ø–∞—Ä—ã: —ç–ª–µ–º–µ–Ω—Ç –∏  –ø—É—Ç—å
    {
        QGraphicsTextItem *textItem = scene->addText(sea[i].second);
        textItem->setFont(QFont("Cascadia Code", 20, 2, false));
        textItem->setPos(-200, y-100);

        QGraphicsTextItem *textItem1 = scene->addText(sea[i].first);
        textItem1->setFont(QFont("Cascadia Code", 20, 2, false));
        textItem1->setPos(300, y-100);
        y-=100;
    }
    QGraphicsTextItem *textItem = scene->addText("–†–µ–∑—É–ª—å—Ç–∞—Ç—ã –ø–æ–∏—Å–∫–∞:"); // –ø–µ—á–∞—Ç—å –∑–∞–≥–æ–ª–æ–≤–∫–∞
    textItem->setFont(QFont("Cascadia Code", 20, 2, false));
    textItem->setPos(-140, y-200);

    QString result = "–ù–∞–π–¥–µ–Ω–æ "; // ... –∏ –∫–æ–ª-–≤–∞ –Ω–∞–π–¥–µ–Ω–Ω—ã—Ö —ç–ª–µ–º–µ–Ω—Ç–æ–≤
    result.push_back(QString::number(sea.size()));
    result.push_back(" —ç–ª–µ–º–µ–Ω—Ç–∞(–æ–≤).");
    QGraphicsTextItem *textItem1 = scene->addText(result);
    textItem1->setFont(QFont("Cascadia Code", 20, 2, false));
    textItem1->setPos(-150, y-100);

    ui->woodTV->update();
    sea.clear();

    ui->findElemsBTN->setHidden(true); // –∫–Ω–æ–ø–∫–∏ –º–µ–Ω—è–µ–º
    ui->closeSearchBTN->setHidden(false);
}

void BinaryWoods::closeSearch() // –∑–∞–∫—Ä—ã—Ç—å –ø–æ–∏—Å–∫ –∏ —É–±—Ä–∞—Ç—å —Å —Ç–µ–ª–µ–∫–∞ –ª–∏—à–Ω–µ–µ
{
    ui->findElemsBTN->setHidden(false);
    ui->closeSearchBTN->setHidden(true);
    reCreateTV();
    printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood);
    allBTNsUnBlock();
}

void BinaryWoods::delWood() // —É–¥–∞–ª—è–µ—Ç –≤—Å–µ –¥–µ—Ä–µ–≤–æ
{
    if (binaryWood!=nullptr)
    {
        binaryWood->~Wood();
        binaryWood = nullptr;
    }
    allBTNsBlock();
    ui->createWoodBTN->setText("–°–æ–∑–¥–∞—Ç—å –¥–µ—Ä–µ–≤–æ");
    ui->createWoodBTN->setEnabled(true);
    reCreateTV();
}

void BinaryWoods::obhodPrinter() // –ø–µ—á–∞—Ç–∞–µ—Ç —Ä–µ–∑—É–ª—å—Ç–∞—Ç –æ–±—Ö–æ–¥–∞ –∏–∑ –≥–ª–æ–±–∞–ª—å–Ω–æ–π —Å—Ç—Ä–æ–∫–∏
{
    int y = -(obh.size()+1)*50;
    for(unsigned int i = 0; i < obh.size(); i++)
    {
        QGraphicsTextItem *textItem = scene->addText(obh[i]);
        textItem->setFont(QFont("Cascadia Code", 20, 2, false));
        textItem->setPos(-50, y+50);
        y+=50;
    }
    ui->woodTV->update();
}

void direct(Wood *tree) //–ø—Ä—è–º–æ–π –æ–±—Ö–æ–¥, —Å–æ–∑—Ä–∞–Ω—è–µ—Ç –≤ –≤–µ–∫—Ç–æ—Ä —Å –ø–∞—Ä–∞–º–∏
{
    if (tree!=NULL) //–ü–æ–∫–∞ –Ω–µ –≤—Å—Ç—Ä–µ—Ç–∏—Ç—Å—è –ø—É—Å—Ç–æ–π —É–∑–µ–ª
    {
        QString tmp;
        for (int i = 0; i < tree->dataSize; i++) tmp.push_back(tree->data[i]);
        obh.push_back(tmp);
        direct(tree->left); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ª–µ–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
        direct(tree->right); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ø—Ä–∞–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
    }
}

void BinaryWoods::directObhod() // –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç –ø—Ä—è–º–æ–π –æ–±—Ö–æ–¥ –∏ –≤—ã–≤–æ–¥–∏—Ç —Ä–µ–∑—É–ª—å—Ç–∞—Ç
{
    allBTNsBlock();
    obh.clear();
    direct(binaryWood);
    obhodPrinter();
    int y = -(obh.size()+2)*50;
    QGraphicsTextItem *textItem = scene->addText("–ü—Ä—è–º–æ–π –æ–±—Ö–æ–¥:");
    textItem->setFont(QFont("Cascadia Code", 20, 2, false));
    textItem->setPos(-50, y+50);
    obh.clear();
    ui->closeObhodBTN->setHidden(false);
    ui->closeObhodBTN->setGeometry(10, 430, 291, 51);
}

void symmetrical(Wood *tree) //—Å–∏–º–º–µ—Ç—Ä–∏—á–Ω—ã–π –æ–±—Ö–æ–¥
{
    if (tree!=NULL) //–ü–æ–∫–∞ –Ω–µ –≤—Å—Ç—Ä–µ—Ç–∏—Ç—Å—è –ø—É—Å—Ç–æ–π —É–∑–µ–ª
    {
      symmetrical(tree->left); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ª–µ–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
      QString tmp;
      for (int i = 0; i < tree->dataSize; i++) tmp.push_back(tree->data[i]);
      obh.push_back(tmp);
      symmetrical(tree->right); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ø—Ä–∞–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
    }
}

void BinaryWoods::symmetricalObhod() // –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç —Å–∏–º–º–µ—Ç—Ä–∏—á–Ω—ã–π –æ–±—Ö–æ–¥ –∏ –≤—ã–≤–æ–¥–∏—Ç –Ω–∞ —Ç–µ–ª–µ–∫
{
    allBTNsBlock();
    obh.clear();
    symmetrical(binaryWood);
    obhodPrinter();
    int y = -(obh.size()+2)*50;
    QGraphicsTextItem *textItem = scene->addText("–°–∏–º–º–µ—Ç—Ä. –æ–±—Ö–æ–¥:");
    textItem->setFont(QFont("Cascadia Code", 20, 2, false));
    textItem->setPos(-50, y+50);
    obh.clear();
    ui->closeObhodBTN->setHidden(false);
    ui->closeObhodBTN->setGeometry(10, 500, 291, 51);
}

void reverse(Wood *tree) //–æ–±—Ä–∞—Ç–Ω—ã–π –æ–±—Ö–æ–¥, —Å–æ–∑—Ä–∞–Ω—è–µ—Ç –≤ –≤–µ–∫—Ç–æ—Ä —Å –ø–∞—Ä–∞–º–∏
{
    if (tree!=NULL) //–ü–æ–∫–∞ –Ω–µ –≤—Å—Ç—Ä–µ—Ç–∏—Ç—Å—è –ø—É—Å—Ç–æ–π —É–∑–µ–ª
    {
        direct(tree->left); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ª–µ–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤–∞
        direct(tree->right); //–†–µ–∫—É—Ä—Å–∏–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –ø—Ä–∞–≤–æ–≥–æ –ø–æ–¥–¥–µ—Ä–µ–≤a
        QString tmp;
        for (int i = 0; i < tree->dataSize; i++) tmp.push_back(tree->data[i]);
        obh.push_back(tmp);
    }
}

void BinaryWoods::reverseObhod() // –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç –æ–±—Ä–∞—Ç–Ω—ã–π –æ–±—Ö–æ–¥ –∏ –≤—ã–≤–æ–¥–∏—Ç –Ω–∞ —Ç–µ–ª–µ–∫
{
    allBTNsBlock();
    obh.clear();
    reverse(binaryWood);
    obhodPrinter();
    int y = -(obh.size()+2)*50;
    QGraphicsTextItem *textItem = scene->addText("–û–±—Ä–∞—Ç–Ω—ã–π –æ–±—Ö–æ–¥:");
    textItem->setFont(QFont("Cascadia Code", 20, 2, false));
    textItem->setPos(-50, y+50);
    obh.clear();
    ui->closeObhodBTN->setHidden(false);
    ui->closeObhodBTN->setGeometry(10, 570, 291, 51);
}

void BinaryWoods::closeObhod() // —á–∏—Å—Ç–∏—Ç —Ç–µ–ª–µ–∫ –æ—Ç –æ–±—Ö–æ–¥–æ–≤ –∏ –≤–∫–ª—è—á–∞–µ—Ç –Ω—É–∂–Ω—ã–µ –∫–Ω–æ–ø–∫–∏
{
    ui->closeObhodBTN->setHidden(true);
    reCreateTV();
    printWood((pow(2, binaryWood->depth-1)*55), 0, 0, binaryWood);
    allBTNsUnBlock();
}

BinaryWoods::~BinaryWoods()
{
   delete ui;
}

